### PR TITLE
[RFC] Add ArrayMap, ArrayFlapMap and ArrayReduce operations

### PR DESCRIPTION
**Pull request by magiblot**
`ArrayMap`, `ArrayFlatMap` and `ArrayReduce` are elemental higher-order array operators. I got inspiration from the similarly-named functions in JavaScript:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

### Tested behaviour

* This script:
  ```py
  BlankClip(pixel_type="YV12")
  [1, 2, 3, 4, 5, 6, 7, 8, 9]
  \   .ArrayMap(function(int i) { [String(i), i] })
  \   .ArrayReduce(last, function(clip c, val v) { c.Subtitle(v[0], align=v[1]) })
  ```
  Produces the following output:
  mpv-shot0001
* This other script:
  ```py
  text = [["a", [10, 20, 30]], ["b", [40, 50, 60]]]
  \   .ArrayFlatMap(function(val v) {
          v[1].ArrayMap(function[v](int i) {
              v[0]+String(i)
          })
      })
  \   .ArrayReduce(function(string acc, string s) {
          acc+", "+s
      })

  BlankClip(pixel_type="YV12")
  Subtitle(text)
  ```
  Produces the following output:
  mpv-shot0001
* Empty arrays are handled properly:
  ```py
  Assert([].ArrayMap(function() {}).ArraySize == 0)
  Assert([].ArrayFlatMap(function() {}).ArraySize == 0)
  Assert([].ArrayReduce($AAAA, function() {}) == $AAAA)
  [].ArrayReduce(function() {}) # ArrayReduce: cannot reduce empty array with no initial value.
  ```
* Basic error cases are handled properly:
  ```py
  "a".ArrayMap(function() {}) # ArrayMap: first parameter must be an array.
  [].ArrayMap(1) # Script error: Invalid arguments to function 'ArrayMap'.
  [1].ArrayMap(function() {}) # ArrayMap: cannot process array element 0: invalid arguments to callback function.
  [1, ""].ArrayReduce(1, function(int acc, int elem) { acc }) # ArrayReduce: cannot process array element 1: invalid arguments to callback function.
  [1].ArrayReduce(1, function(int acc, int i) { Assert(false) }) # Assert: assertion failed
  [1].ArrayFlatMap(function(string s) {}) # ArrayFlatMap: cannot process array element 0: invalid arguments to callback function.
  [1].ArrayFlatMap(function(int i) { i }) # ArrayFlatMap: cannot process array element 0: callback function must return an array (got 'int').
  ```
### Issues

<details>
<summary><s>Missing MTGuards</s></summary>

I'm probably doing something wrong, because calling function objects from `ArrayMap` or `ArrayReduce` results in `MTGuard`s being missed.

For example, when I run the following script and set a breakpoint on `SingleFrame::GetFrame`:
```py
f = function(a, b) {
  BlankClip(pixel_type="YV12")
  AutoLevels
  AutoLevels
  AutoLevels
}

f(1, 2)
```

I get the following backtrace:

```c
#0  SingleFrame::GetFrame
#1  Autolevels::getStats
#2  Autolevels::isSceneStart
#3  Autolevels::GetFrame
#4  MTGuard::GetFrame /* OK! */
#5  Cache::GetFrame
#6  CacheGuard::GetFrame
#7  Autolevels::getStats
#8  Autolevels::isSceneStart
#9  Autolevels::GetFrame
#10 MTGuard::GetFrame /* OK! */
#11 Cache::GetFrame
#12 CacheGuard::GetFrame
#13 Autolevels::getStats
#14 Autolevels::isSceneStart
#15 Autolevels::GetFrame
#16 MTGuard::GetFrame /* OK! */
#17 Cache::GetFrame
#18 CacheGuard::GetFrame
#19 Prefetcher::GetFrame
```

But when I run this other script:
```py
f = function(a, b) {
  BlankClip(pixel_type="YV12")
  AutoLevels
  AutoLevels
  AutoLevels
}

[1].ArrayReduce(1, f)
```

The backtrace turns out like this:
```c
#0  SingleFrame::GetFrame
#1  Autolevels::getStats
#2  Autolevels::isSceneStart
#3  Autolevels::GetFrame /* No MTGuard? */
#4  Cache::GetFrame
#5  CacheGuard::GetFrame
#6  Autolevels::getStats
#7  Autolevels::isSceneStart
#8  Autolevels::GetFrame /* No MTGuard? */
#9  Cache::GetFrame
#10 CacheGuard::GetFrame
#11 Autolevels::getStats
#12 Autolevels::isSceneStart
#13 Autolevels::GetFrame /* No MTGuard? */
#14 Cache::GetFrame
#15 CacheGuard::GetFrame /* What? */
#16 MTGuard::GetFrame
#17 Cache::GetFrame
#18 CacheGuard::GetFrame
#19 Prefetcher::GetFrame
```
</details>

Thank you.

### CODE REVIEW

**Comment by pinterf** [comment]
Thanks, ArrayMap is basically a foreach with lambda?
As for no MTGuard seen: `PClip MTGuard::Create` only creates the object when it is not MT_NICE_FILTER. AutoLevels is which kind of a filter?

**Comment by magiblot** [comment]
Hi Ferenc,

`map` and `reduce` (among others) are commonly-used functions in functional programming. Yes, ArrayMap is similar to a foreach except that it is meant to generate a new list and not modify the original one.

`AutoLevels` does not specify a MT mode, so I guess it's MT_MULTI_INSTANCE?

**Comment by magiblot** [comment]
Okay, I've found the reason why MTGuards are not being created:

https://github.com/AviSynth/AviSynthPlus/blob/efcd254f6fc4dfc8d87b8e14c25a59a4e32f10ee/avs_core/core/avisynth.cpp#L4545-L4548

I'm not sure how to proceed in this case...

**Comment by pinterf** [comment]
What can you see, when the function is called in runtime environment, from ScriptClip which is doing the same task which is in your example?

**Comment by magiblot** [comment]
With `ScriptClip` there are no `MTGuard`s either. However, after taking a closer look at `ScriptEnvironment::Invoke_` I've found a way to do it. There are two cases where it works:
* If the MT mode of the filter is `MT_SPECIAL_MT`: https://github.com/AviSynth/AviSynthPlus/blob/efcd254f6fc4dfc8d87b8e14c25a59a4e32f10ee/avs_core/core/avisynth.cpp#L4599
* If the filter is considered a *script function* (like `Eval`, `Import`...): https://github.com/AviSynth/AviSynthPlus/blob/efcd254f6fc4dfc8d87b8e14c25a59a4e32f10ee/avs_core/core/avisynth.cpp#L4606

To make `ArrayMap` and `ArrayReduce` *script functions*, one has to add them to `AVSFunction::IsScriptFunction`. To make them `MT_SPECIAL_MT`, it is necessary to invoke `env->SetFilterMTMode` at some point.

I wonder which of these options is preferable?

**Comment by pinterf** [comment]
I'd avoid SPECIAL_MT, it was just an unofficial workaround for ScriptSclip and other MT deadlock and threading problems, which were healed after backporting Neo fork's fixes. I'm gonna have a look at it next week.
Is your target to have function objects to perform complex filter operation with adjustable MT behaviour inside?

**Comment by pinterf** [comment]
The IsScriptFunction section contains a comment
`// Eval, EvalOop, Import and user defined script functions`
Investigating the history for PluginManager script functions were added to the list here: 2def848f
I guess function objects are not recognized by this line because they are not called directly from the script flow, but 'manually' from within these two new array-support function?

**Comment by magiblot** [comment]
The problem are not function objects, but the new filters (`ArrayMap`, `ArrayReduce`). When these filters invoke the function object,  the `chainedCtor` condition is true, so filters instantiated within the function object do not get `MTGuard`s. If `MT_SPECIAL_MT` is to be avoided, the only way I see to avoid this condition is to add the new filters to `AVSFunction::IsScriptFunction`.

To put it another way, `ArrayMap` and `ArrayReduce` should just behave like they transform a piece of code into another one. For example, the following piece of code:

```py
f = function(a, b) {
  BlankClip(pixel_type="YV12")
  AutoLevels
  AutoLevels
  AutoLevels
}

[1, 2, 3].ArrayMap(f)
```

Should behave exactly the same as this other piece of code:

```py
f = function(a, b) {
  BlankClip(pixel_type="YV12")
  AutoLevels
  AutoLevels
  AutoLevels
}

[f(1), f(2), f(3)]
```

And in this sense `ArrayMap` and `ArrayReduce` are similar to filters like `Eval`.

**Comment by pinterf** [comment]
It makes sense to expand the list with these two new functions. They behave like a macro expansion but this cannot be recognized easily.

**Comment by qyot27** [comment]
Somewhat tangential: as this would be adding new script functions, this would almost certainly meet the criteria for bumping `AVISYNTHPLUS_INTERFACE_BUGFIX_VERSION`.

**Comment by magiblot** [comment]
Thanks @qyot27, I take note of that.

**Comment by pinterf** [comment]
Yes, we must step the bugfix version to the next one immediately after a release to have a different number from the official one. e.g. 3.7.2. I build the unofficial test versions with this version number. (like it was done with my 3.7.1 test series)

Then this version number is either kept for the release (as now) or we make a distinct next version just for the having a unique number for the release.

**Comment by pinterf** [comment]
(But versioning is not the scope of this PR, qyot27 or I will do that if the proposal above is acceptable)

**Comment by magiblot** [comment]
Ah, okay. Thank you :)

**Comment by qyot27** [comment]
>Yes, we must step the bugfix version to the next one immediately after a release to have a different number from the official one. e.g. 3.7.2. I build the unofficial test versions with this version number. (like it was done with my 3.7.1 test series)

`AVISYNTHPLUS_INTERFACE_BUGFIX_VERSION` can be bumped whenever new filters, changes to filter behavior that need checks, or API fixes get added, as it doesn't affect the release number.  Instead of being 0, the addition of ArrayMap/Reduce would bump it to 1.  It only gets reset to zero once the value of `AVISYNTH_INTERFACE_VERSION` gets bumped, like what happened shortly before 3.7.1 was released (interface version 8.1 became interface version 9.0).  There's a much greater likelihood that when we would eventually bump to interface version 10, that the bugfix value for interface 9 could be in the 20s, 30s, or possibly even higher - it just depends on how much gets fixed or added and how deeply (without affecting the base API).

In short,
- `AVISYNTH_INTERFACE_VERSION` denotes changes that break compatibility with plugins and clients
- `AVISYNTHPLUS_INTERFACE_BUGFIX_VERSION` denotes changes that don't break compatibility (or fix compatibility for things that weren't actually regressions), or when adding things like new filters, so that it can be quickly looked up by querying said value.  Think of it like a feature set index.  The 'BUGFIX' naming is because the issue that immediately led to its creation was a bugfix.

We probably need a page in the documentation that provides an equivalency chart for this, separate from the changelog so that it doesn't get lost or buried in all the other things the changelog is used for.

**Comment by pinterf** [comment]
Sorry, I mismatched released numbering (3.7.x) with our interface version numbering scheme.

**Comment by magiblot** [comment]
> It makes sense to expand the list with these two new functions. They behave like a macro expansion but this cannot be recognized easily.

Done. In the meantime I also added a third operation (`ArrayFlatMap`).

This is now ready for review.

**Comment by pinterf** [comment]
No time yet other than inspecting visually the patch, they seems good. Until then some question; I'm curious because you made these additions for purpose.
Did these new features make your script easier to read or maintain? What is your use case for them, is it that big MT-problematic script? What did the IsScriptFunction MT guard additions mean for you performance or memory-wise?

**Comment by magiblot** [comment]
Yes, I submitted these changes because they should help me with the script from #244. For example, that script contains several lines like these:

```py
ReplaceStabCrop(b, 5, 20, 45, 559, 136, "[11041 11122] [11505 11564] [11590 11660]")
ReplaceStabCrop(b, 5, 678, 53, 16, 65, "[11879 11917]")
ReplaceStabCrop(b, 5, 644, 47, 18, 59, "[13617 13755]")
ReplaceStabCrop(b, 5, 41, 58, 504, 247, "[13803 14111] [14222 14314]")
```

In this case, `ReplaceStabCrop` is meant to apply stabilization to one or more ranges of frames. This involves instantiating `DePanEstimate` and other filters.

A difficulty of this is to prevent frames not in these ranges from interfering in the stabilization process. Originally I was replacing out-of-range frames with blank ones:

```py
function ReplaceStabCrop(clip c, clip blank, int range, int left, int top, int right, int bottom, string mappings) {
    blank.ReplaceFramesSimple(c, mappings=mappings)
    # ...
}
```

But with array manipulators I can programmatically `Trim` the clip before passing it to `DePan`:

```py
function ReplaceStabCrop(clip c, clip blank, int range, int left, int top, int right, int bottom, string mappings) {
    val args = Eval("Array("+mappings.ReplaceStr(" ", ",")+")") # could also just take an array parameter.
    args.ArrayReduce(c, function[blank, range, left, top, right, bottom](acc, v)) {
        begin = v[0]
        end = v[1]
        trimmed = acc.Trim(begin, end=end)
        stabbed = StabCrop(trimmed, range, left, top, right, bottom) # Calls Crop, DePan, etc.
        acc.RemapFrames(sourceClip=stabbed, mappings="[{} {}] [0 {}]".Format(begin, end, end - begin))
    })
}
```

This may lead to memory savings because some buffers allocated by `Depan`, `DePanEstimate` and other filters are proportionally large to the length of its input clip. Further memory could be saved if `RemapFrames` took an array of source clips as input so that only one instance of it is necessary. These savings are just theoretical, I haven't put this into practice yet.

Although the new array operators are not strictly necessary to do this, they allow for cleaner and more maintainable code.

Another possible advantage is the ability to reduce code verbosity. For example, `ArrayReduce` makes it possible to turn all the `ReplaceStabCrop` invocations into a single one, reducing the amount of code that has to be typed each time:

```py
function ReplaceStabCropMany(clip c, clip b, int range, val args) {
    args.ArrayReduce(c, function[b, range](clip acc, val v) {
        acc.ReplaceStabCrop(b, range, v[0], v[1], v[2], v[3], v[4])
    })
}

ReplaceStabCropMany(b, 5, [
\   [20, 45, 559, 136, "[11041 11122] [11505 11564] [11590 11660]"],
\   [678, 53, 16, 65, "[11879 11917]"],
\   [644, 47, 18, 59, "[13617 13755]"],
\   [41, 58, 504, 247, "[13803 14111] [14222 14314]"]
\   ])
```

Cheers.

**Comment by magiblot** [comment]
I can confirm this was useful. At 1080p resolution and with more `DePan` instances than in the script I shared with you, memory usage went from 8.19 GB to 5.91 GB after using the new array functions in the way I mentioned.

**Comment by magiblot** [comment]
I just realized it may be more convenient to simply define these functions directly in the AviSynth language...

```js
function ArrayMap(val v, func f) {
    Assert(v.IsArray, "ArrayMap: first argument must be an array")
    a = []
    for (i = 0, v.ArraySize - 1) {
        a = a.ArrayAdd(f(v[i]))
    }
    a
}

function ArrayFlatMap(val v, func f) {
    Assert(v.IsArray, "ArrayFlatMap: first argument must be an array")
    a = []
    for (i = 0, v.ArraySize - 1) {
        b = f(v[i])
        Assert(b.IsArray, "ArrayFlatMap: cannot process array element {}: callback function must return an array (got '{}')".Format(i, b.TypeName))
        for (j = 0, b.ArraySize - 1) {
            a = a.ArrayAdd(b[j])
        }
    }
    a
}

function ArrayReduce(val v, func f, val "initial") {
    Assert(v.IsArray, "ArrayReduce: first argument must be an array")
    if (Defined(initial)) {
        i = 0
        acc = initial
    } else {
        Assert(v.ArraySize > 0, "ArrayReduce: cannot reduce empty array with no initial value")
        i = 1
        acc = v[0]
    }
    for (i = i, v.ArraySize - 1) {
        acc = f(acc, v[i])
    }
    acc
}
```

Considering how easy it is to write them in Avisynth and the trouble it takes to implement them in C++, changing the AviSynth interface does not seem justified. It's probably best to dismiss this PR.

**Comment by magiblot** [comment]
Closing. Thanks for the attention!

**Comment by pinterf** [comment]
It depends. Maybe MT and thus memory or performancewise the c++ implementation can be better controlled? Plus the convenience that everyone can access them w/o including additional avsi stuff. Do the script versions behave the very same way?
